azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/async_internal_client.py

```diff
@@ -0,0 +1,95 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+import logging
+import asyncio
+import functools
+from .internal_client import InternalClient
+
+logger = logging.getLogger(__name__)
+
+
+# all of our async wrappers can follow this same pattern.  We might be able to make this into a
+# helper function, but that will involve some tricky parameter handling because the signature of
+# callback changes depending on the function.
+
+# these are currently just functions in the module namespace.  We add them to the InternalClient
+# prototype below inside the monkeypatch() function.
+
+
+async def _connect_async(self):
+    """
+    This is just a test docstring to make sure we can see documentation for these functions
+    when we call help(InternalClient), and we can, so life is good and everyone is happy.
+    """
+    logger.info("async connecting to transport")
+
+    # we need the event loop so we can run our code in side it.
+    loop = asyncio.get_event_loop()
```

Comment:
>loop = asyncio.get_event_loop() [](start = 4, length = 31)

What happens when we use `get_running_loop()` ?

Reply:
okay. it is written in 3.7 as to being preferable to use. The reason being cited is `Because this function has rather complex behavior (especially when custom event loop policies are in use), using the get_running_loop() function is preferred to get_event_loop() in coroutines`. I am thinking this one does not return the `currently running event loop` but returns the current thread bound one though I am not exactly sure of the implication. 

---
In reply to: [238760013](https://github.com/Azure/azure-iot-sdk-python-preview/pull/28#discussion_r238760013) [](ancestors = 238760013)

Reply:
That's really good info.  I'm reading that it works better in complex situations, so maybe we should have the code use get_running_loop if it can, and fallback to get_event_loop.  Hopefully people with complex async/await setups are using 3.7, so this would satisfy them while still working downlevel.

Reply:
suggestion: typically a good question to ask on the Python teams channel

---
In reply to: [238881039](https://github.com/Azure/azure-iot-sdk-python-preview/pull/28#discussion_r238881039) [](ancestors = 238881039)